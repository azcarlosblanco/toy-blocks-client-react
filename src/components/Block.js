import React from "react";
import PropTypes from "prop-types";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import {
  ExpansionPanel,
  ExpansionPanelSummary,
  Typography,
  ExpansionPanelDetails,
  makeStyles,
  Box,
  Grid,
} from "@material-ui/core";
import colors from "../constants/colors";
import Status from "./Status";
import Blocks from "../containers/Blocks";

const Block = ({ item: {attributes}  }) => {
  const classes = useStyles();

  return (
    <Grid className={classes.grid} item xs={12}>
      <Typography className={classes.text1}>{('000' + attributes.index).substr(-3)}</Typography>
      <Typography className={classes.text2}>{attributes.data}</Typography>
    </Grid>
  );
};

const useStyles = makeStyles((theme) => ({
  grid: {
    margin: "5px 0",
    padding: "5px",
    backgroundColor: "#CCCCCC",
    borderRadius: "5px"
  },
  text1: {
    color: "blue",
    fontSize: '10px',
    fontWeight: "bold"
  },
  text2: {
    color: "black",
  },
  
}));

Block.propTypes = {
  item: PropTypes.object.isRequired
};

export default Block;
