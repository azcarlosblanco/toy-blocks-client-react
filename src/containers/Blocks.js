import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import * as actions from "../actions/nodes";
import Node from "../components/Node";
import { Typography, Box, CircularProgress } from "@material-ui/core";
import Block from "../components/Block";

export class Blocks extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      blocks: [],
      loading: true,
      error: false,
    };
  }

  async componentDidMount() {
    const data = await this.getBlocks();
    this.setState({...data, loading: false})
  }

  async getBlocks() {
    try {
      const node = await fetch(`${this.props.nodeUrl}/api/v1/blocks`)
      const nodeResponse = await node.json();

      if (node.status === 200) {
        return {
          blocks: nodeResponse.data,
          error: false
        }
      }

      return {
        blocks: [],
        error: true
      }
    } catch (error) {
      return {
        blocks: [],
        error: true
      }
    }
  }

  renderBlocks() {
    debugger
    return this.state.blocks.map((item, i) => <Block key={i} item={item} />);
  }

  render() {
    if (this.state.loading === true) {
      return <CircularProgress />
    }

    if (this.state.error === true) {
      return <Typography variant="h5">Something went wrong!!!</Typography>;
    }

    if (this.state.blocks.length === 0) {
      return <Typography variant="h5">No blocks were received.</Typography>;
    }

    return this.renderBlocks();
  }
}

Blocks.propTypes = {
  nodeUrl: PropTypes.string.isRequired,
};

export default Blocks;
