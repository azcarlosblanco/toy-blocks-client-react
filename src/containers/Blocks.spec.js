import React from "react";
import { mount, shallow } from "enzyme";
import configureMockStore from "redux-mock-store";
import thunk from 'redux-thunk';
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import ConnectedNodes, { Nodes } from "./Nodes";
import Node from "../components/Node";
import { CircularProgress, Typography } from "@material-ui/core";
import Blocks from "./Blocks";
import Block from "../components/Block";

describe("<Blocks />", () => {
    const nodeUrl = 'http://testUrl';
    const blocks = [
        {
            "id": "5",
            "type": "blocks",
            "attributes": {
              "index": 1,
              "timestamp": 1530679678,
              "data": "The Human Torch",
              "previous-hash": "KsmmdGrKVDr43/OYlM/oFzr7oh6wHG+uM9UpRyIoVe8=",
              "hash": "oHkxOJWOKy02vA9r4iRHVqTgqT+Afc6OYFcNYzyhGEc="
            }
        }
    ]

  it("should render a loader", async () => {
      const wrapper = shallow(<Blocks nodeUrl={nodeUrl} />);
      const loader = wrapper.find(CircularProgress).exists();
      expect(loader).toBeTruthy();
  });

  it("should render an error", async () => {
    jest.spyOn(Blocks.prototype, 'getBlocks').mockReturnValue({
        error: true,
        blocks: []
    });

    const wrapper = await shallow(<Blocks nodeUrl={nodeUrl} />);
    const textError = await wrapper.find(Typography).text();
    expect(textError).toEqual('Something went wrong!!!');
  });

  it("should render an empty state", async () => {
    jest.spyOn(Blocks.prototype, 'getBlocks').mockReturnValue({
        error: false,
        blocks: []
    });

    const wrapper = await shallow(<Blocks nodeUrl={nodeUrl} />);
    const textError = await wrapper.find(Typography).text();
    expect(textError).toEqual('No blocks were received.');
  });

  // Blocks getBlocks() that is going to return an object with blocks on it
  it("should render blocks", async () => {
    jest.spyOn(Blocks.prototype, 'getBlocks').mockReturnValue({
        error: false,
        blocks
    });

    const wrapper = await mount(<Blocks nodeUrl={nodeUrl} />);
    wrapper.update()

    const text1 = wrapper.find(Block).at(0).find(Typography).at(0).props();
    const text2 = wrapper.find(Block).at(0).find(Typography).at(1).props();

    expect(text1.children).toEqual('001');
    expect(text2.children).toEqual('The Human Torch');
  });

  it("should match snapshot", async () => {
    jest.spyOn(Blocks.prototype, 'getBlocks').mockReturnValue({
        error: false,
        blocks
    });

    const component = await create(<Blocks nodeUrl={nodeUrl} />);
    const tree = component.toJSON();
    
    expect(tree).toMatchSnapshot();
  });
});
