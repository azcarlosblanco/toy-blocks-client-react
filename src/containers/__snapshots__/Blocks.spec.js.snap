// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blocks /> should match snapshot 1`] = `
<div
  className="MuiGrid-root makeStyles-grid-13 MuiGrid-item MuiGrid-grid-xs-12"
>
  <p
    className="MuiTypography-root makeStyles-text1-14 MuiTypography-body1"
  >
    001
  </p>
  <p
    className="MuiTypography-root makeStyles-text2-15 MuiTypography-body1"
  >
    The Human Torch
  </p>
</div>
`;
